// @react-dev-inspector/babel-plugin@2.0.1/es/transform.js
import path from 'path';
import { parse, } from '@babel/parser';
import generate from '@babel/generator';
import traverse from '@babel/traverse';
import { pathMatch, doJSXOpeningElement, } from './visitor';
/**
 * inject line, column, relative-path to JSX html data attribute in source code
 *
 * @type webpack.loader.Loader
 * ref: https://astexplorer.net  +  @babel/parser
 */
export const transform = ({ rootPath = process.cwd(), filePath, sourceCode, options, }) => {
    var _a;
    /**
     * example:
     * rootPath: /home/xxx/project
     * filePath: /home/xxx/project/src/ooo/xxx.js
     * relativePath: src/ooo/xxx.js
     */
    const relativePath = path.relative(rootPath, filePath);
    const isSkip = pathMatch(filePath, options === null || options === void 0 ? void 0 : options.excludes);
    if (isSkip) {
        return sourceCode;
    }
    const ast = parse(sourceCode, {
        sourceType: 'module',
        allowUndeclaredExports: true,
        allowImportExportEverywhere: true,
        plugins: [
            'typescript',
            'jsx',
            'decorators-legacy',
            'classProperties',
            ...(_a = options === null || options === void 0 ? void 0 : options.babelPlugins) !== null && _a !== void 0 ? _a : [],
        ],
        ...options === null || options === void 0 ? void 0 : options.babelOptions,
    });
    /**
     * astexplorer + @babel/parser
     * https://astexplorer.net
     */
    traverse(ast, {
        JSXOpeningElement: {
            enter(path) {
                doJSXOpeningElement(path.node, { relativePath });
            },
        },
    });
    const { code, } = generate(ast, {
        decoratorsBeforeExport: true,
    });
    return code;
};
