// react-dev-inspector@2.0.1/lib/Inspector/utils/overlay.js
"use strict";
/**
 * mirror from https://github.com/facebook/react/blob/v16.13.1/packages/react-devtools-shared/src/backend/views/Highlighter/Overlay.js
 *
 * remove all process for iframe, because iframe only need to think in chrome extension app,
 * which will deal multiple levels of nesting iframe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getElementDimensions = exports.getNestedBoundingClientRect = void 0;
function getNestedBoundingClientRect(node) {
    return node.getBoundingClientRect();
}
exports.getNestedBoundingClientRect = getNestedBoundingClientRect;
function getElementDimensions(domElement) {
    const calculatedStyle = window.getComputedStyle(domElement);
    return {
        borderLeft: Number.parseInt(calculatedStyle.borderLeftWidth, 10),
        borderRight: Number.parseInt(calculatedStyle.borderRightWidth, 10),
        borderTop: Number.parseInt(calculatedStyle.borderTopWidth, 10),
        borderBottom: Number.parseInt(calculatedStyle.borderBottomWidth, 10),
        marginLeft: Number.parseInt(calculatedStyle.marginLeft, 10),
        marginRight: Number.parseInt(calculatedStyle.marginRight, 10),
        marginTop: Number.parseInt(calculatedStyle.marginTop, 10),
        marginBottom: Number.parseInt(calculatedStyle.marginBottom, 10),
        paddingLeft: Number.parseInt(calculatedStyle.paddingLeft, 10),
        paddingRight: Number.parseInt(calculatedStyle.paddingRight, 10),
        paddingTop: Number.parseInt(calculatedStyle.paddingTop, 10),
        paddingBottom: Number.parseInt(calculatedStyle.paddingBottom, 10),
    };
}
exports.getElementDimensions = getElementDimensions;
