// @react-dev-inspector/umi4-plugin@2.0.1/es/index.js
/**
 * preset plugins for umi4
 */
import { createLaunchEditorMiddleware, } from '@react-dev-inspector/middleware';
export default function inspectorPlugin(api) {
    var _a, _b;
    const inspectorConfig = api.userConfig.inspectorConfig;
    api.describe({
        key: 'inspectorConfig',
        config: {
            schema(joi) {
                return joi.object();
            },
        },
        enableBy: api.EnableBy.register,
    });
    const enableBabelPlugin = (_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.enableBabelPlugin) !== null && _a !== void 0 ? _a : (process.env.NODE_ENV === 'development'
        && ((_b = api.config.srcTranspiler) !== null && _b !== void 0 ? _b : 'babel') === 'babel');
    if (enableBabelPlugin) {
        // https://umijs.org/docs/api/plugin-api#addbeforebabelplugins
        api.addBeforeBabelPlugins(() => {
            var _a;
            return [
                [
                    require.resolve('@react-dev-inspector/babel-plugin'),
                    {
                        cwd: inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.cwd,
                        excludes: [
                            /\.umi(-production)?\//,
                            ...(_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.excludes) !== null && _a !== void 0 ? _a : [],
                        ],
                    },
                ],
            ];
        });
    }
    // https://umijs.org/docs/api/plugin-api#addbeforemiddlewares
    api.addBeforeMiddlewares(createLaunchEditorMiddleware);
}
