// @react-dev-inspector/umi3-plugin@2.0.1/lib/index.js
"use strict";
/**
 * preset plugins for umi3
 */
Object.defineProperty(exports, "__esModule", { value: true });
const middleware_1 = require("@react-dev-inspector/middleware");
function inspectorPlugin(api) {
    var _a;
    const inspectorConfig = api.userConfig.inspectorConfig;
    api.describe({
        key: 'inspectorConfig',
        config: {
            schema(joi) {
                return joi.object();
            },
        },
        enableBy: api.EnableBy.register,
    });
    const enableBabelPlugin = (_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.enableBabelPlugin) !== null && _a !== void 0 ? _a : (process.env.NODE_ENV === 'development');
    if (enableBabelPlugin) {
        api.modifyBabelOpts((babelOptions) => {
            var _a;
            babelOptions.plugins.unshift([
                require.resolve('@react-dev-inspector/babel-plugin'),
                {
                    cwd: inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.cwd,
                    excludes: [
                        /\.umi(-production)?\//,
                        ...(_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.excludes) !== null && _a !== void 0 ? _a : [],
                    ],
                },
            ]);
            return babelOptions;
        });
    }
    const createMiddleware = () => middleware_1.launchEditorMiddleware;
    /**
     * due to umi3 not use webpack devServer,
     * so need add launch editor middleware manually
     */
    api.addBeforeMiddlewares
        ? api.addBeforeMiddlewares(createMiddleware)
        // typo of umi3 legacy api
        : api.addBeforeMiddewares(createMiddleware);
}
exports.default = inspectorPlugin;
