// @react-dev-inspector/umi4-plugin@2.0.1/lib/index.js
"use strict";
/**
 * preset plugins for umi4
 */
Object.defineProperty(exports, "__esModule", { value: true });
const middleware_1 = require("@react-dev-inspector/middleware");
function inspectorPlugin(api) {
    var _a, _b;
    const inspectorConfig = api.userConfig.inspectorConfig;
    api.describe({
        key: 'inspectorConfig',
        config: {
            schema(joi) {
                return joi.object();
            },
        },
        enableBy: api.EnableBy.register,
    });
    const enableBabelPlugin = (_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.enableBabelPlugin) !== null && _a !== void 0 ? _a : (process.env.NODE_ENV === 'development'
        && ((_b = api.config.srcTranspiler) !== null && _b !== void 0 ? _b : 'babel') === 'babel');
    if (enableBabelPlugin) {
        // https://umijs.org/docs/api/plugin-api#addbeforebabelplugins
        api.addBeforeBabelPlugins(() => {
            var _a;
            return [
                [
                    require.resolve('@react-dev-inspector/babel-plugin'),
                    {
                        cwd: inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.cwd,
                        excludes: [
                            /\.umi(-production)?\//,
                            ...(_a = inspectorConfig === null || inspectorConfig === void 0 ? void 0 : inspectorConfig.excludes) !== null && _a !== void 0 ? _a : [],
                        ],
                    },
                ],
            ];
        });
    }
    // https://umijs.org/docs/api/plugin-api#addbeforemiddlewares
    api.addBeforeMiddlewares(middleware_1.createLaunchEditorMiddleware);
}
exports.default = inspectorPlugin;
